node_modules/oae-tenants/lib/api.js

```diff
@@ -494,6 +534,64 @@ var _setLandingPageBlockAttribute = function(ctx, block, blockName, attributeNam
 ///////////////
 
 /**
+ * Ensure that the given email domain does not conflict with any existing tenant domains. If this
+ * validation is part of an update, the update tenant alias should be specified so that any conflict
+ * in email domain with the tenant being updated can be ignored. Any validation errors will result
+ * in the errors being added to the provided `validator`, therefore this does not `return` or
+ * `throw` anything.
+ *
+ * @param  {Validator}  validator               The Validator object
+ * @param  {String}     emailDomain             The email domain expression to verify
+ * @param  {String}     [updateTenantAlias]     The alias of the tenant being updated, if any
+ * @api private
+ */
+var _validateEmailDomain = function(validator, emailDomain, updateTenantAlias) {
+    var matchingEmailDomain = _.chain(getTenants())
+        .filter(function(tenant) {
+            // If a tenant is being updated, don't try and match against its own existing email
+            // domain
+            return (tenant.alias !== updateTenantAlias);
+        })
+        .pluck('emailDomain')
+        .compact()
+        .filter(function(emailDomainExpr) {
+            // There is a conflict if there is an existing expression that matches with the
+            // expression, or vice versa
+            return (_matchesEmailDomain(emailDomain, emailDomainExpr) ||
+                _matchesEmailDomain(emailDomainExpr, emailDomain));
+        })
+        .first()
+        .value();
+
+    validator.check(matchingEmailDomain, {
+        'code': 400,
+        'msg': util.format('The email domain "%s" conflicts with existing email domain: %s', emailDomain, matchingEmailDomain)
+    }).isNull();
+};
+
+/**
+ * Determine if the given email domain expression matches the given email domain
+ *
+ * @param  {String}     emailDomain         The email domain to check
+ * @param  {String}     emailDomainExpr     The email domain expression (e.g., *.cam.ac.uk)
```

Comment:
I'm not sure doing smart things like using the longest match is feasible with 1000's of tenants, it seems just too tricky to manage.

I do think now we at least need to avoid having to do `*cam.ac.uk` as the suffix could potentially match `otherstuffcam.ac.uk`, and that could be switched to `*.cam.ac.uk,cam.ac.uk` to support a safer match.